src/idl_parser.cpp

```diff
@@ -912,18 +912,14 @@ CheckedError Parser::ParseField(StructDef &struct_def) {
     FLATBUFFERS_ASSERT(IsInteger(type.base_type) ||
                        (type.base_type == BASE_TYPE_UNION) || IsVector(type) ||
                        IsArray(type));
-    if (IsVector(type)) {
-      // Vector can't use initialization list.
-      FLATBUFFERS_ASSERT(field->value.constant == "0");
-    } else {
-      // All unions should have the NONE ("0") enum value.
-      auto in_enum = field->IsOptional() ||
-                     type.enum_def->attributes.Lookup("bit_flags") ||
-                     type.enum_def->FindByValue(field->value.constant);
-      if (false == in_enum)
-        return Error("default value of " + field->value.constant +
-                     " for field " + name + " is not part of enum " +
-                     type.enum_def->name);
+
```

Comment:
The current `if(IsVector(type))-else` is more readable than proposed solution. The `else` branch covers only scalars or enums.
With the new solution, any high-level error (like = [1,2,3]) will be ignored.

Instead of `FLATBUFFERS_ASSERT` just before `if(IsVector(type))` it might be better to write full decision tree including `IsArray`, for example:
```cpp
if (type.base_type == BASE_TYPE_UNION) {
  if(field->value.constant != "0") return Error("");
} else if (IsInteger(type.base_type)) {
  ...
} else if (IsVector(type)) {
  FLATBUFFERS_ASSERT(field->value.constant == "0" || field->value.constant == "[]");
} else if (IsArray(type)) {
  FLATBUFFERS_ASSERT(field->value.constant == "0");
} else {
  FLATBUFFERS_ASSERT(false && "unexpected type");
}
```

Reply:
Sure, I refactored away the asserts and used a chain of if/else. PTAL

Reply:
Thanks, looks good.

>// Optional and bitflags enums may have default constants that are not
>// their specified variants.

Optional fields might have only `null` initializer to indicate optionality, so it never can be an enum's variant.